torch/csrc/jit/python/pybind_utils.h

```diff
@@ -289,6 +289,43 @@ inline InferredType tryToInferType(py::handle input) {
       TORCH_INTERNAL_ASSERT(class_type);
       return InferredType(class_type);
     }
+
+    // Check if it is a NamedTuple.
```

Comment:
This code is essentially the same as the code you pointed out in `python_sugared_value.cpp`. I tried to reuse that by calling into it, but `python_sugared_value.cpp` includes `pybind_utils.h`. Additionally, it does not seem like `isNamedTupleClass` and `registerNamedTuple` belong in `pybind_utils.h` from an organizational standpoint. The code in `python_sugared_value.cpp` also uses the `SourceRange` for an `ErrorReport` regarding default values not being supported, which I was not able to include here because I wasn't sure how to get the source location info (is it even possible?).

Any suggestions?

Reply:
In general I'm not a fan of non-hot-path code that inspects python objects being in C++. It's much harder to write and debug and doesn't really buy us anything. On the other hand, it's also not great to jump back and forth between python and C++. we're kind of alrreading doing that here with the two `torch.jit` invocations. 

I would personally refactor and move the shared code to python and invoke it from both the functions. either in `torch/jit/__init__.py` or `torch/_jit_internal.py` wherever seems best. that function can take in an optional source range.

Reply:
Had to lift a lot more than I anticipated from C++ -> Python, let me know what you think.